pkg/reconciler/build/build.go

```diff
@@ -152,11 +170,49 @@ func (c *Reconciler) reconcile(ctx context.Context, build *buildapi.Build) error
 
 	build.Status.PodName = pod.Name
 	build.Status.StepStates = stepStates(pod)
-	build.Status.StepsCompleted = stepCompleted(pod)
-	build.Status.Conditions = conditionForPod(pod)
+	build.Status.StepsCompleted = stepsCompleted(pod)
+	build.Status.Conditions = conditionForPod(pod, build.Status.StepsCompleted)
 	return nil
 }
 
+func (c *Reconciler) setBuildReady(ctx context.Context, pod *corev1.Pod) (*corev1.Pod, error) {
+	if _, found := pod.Annotations[buildapi.BuildReadyAnnotation]; found {
+		return pod, nil
+	}
+
+	ready := 0
+	for _, container := range pod.Status.ContainerStatuses {
+		if container.Ready {
+			ready++
+		}
+	}
+	if ready != len(pod.Spec.Containers) {
+		return pod, nil
+	}
+
+	patch, err := json.Marshal([]jsonpatch.JsonPatchOperation{{
+		Operation: "add",
+		Path:      "/metadata/annotations/" + strings.Replace(buildapi.BuildReadyAnnotation, "/", "~1", 1),
+		Value:     "true",
+	}})
+	if err != nil {
+		return nil, err
+	}
+
+	return c.K8sClient.CoreV1().Pods(pod.Namespace).Patch(ctx, pod.Name, types.JSONPatchType, patch, metav1.PatchOptions{})
+}
+
+func (c *Reconciler) cleanupIfNeeded(ctx context.Context, pod *corev1.Pod) (*corev1.Pod, error) {
+	// check if pod is running && completion is terminated
+	if pod.Status.Phase == corev1.PodRunning && contains(stepsCompleted(pod), "completion") {
+		activeDeadlineSeconds := int64(1)
+		podCopy := pod.DeepCopy()
+		podCopy.Spec.ActiveDeadlineSeconds = &activeDeadlineSeconds
+		return c.K8sClient.CoreV1().Pods(pod.Namespace).Update(ctx, podCopy, metav1.UpdateOptions{})
```

Comment:
why update and not patch while the other is patch?

Reply:
honestly no clue, I think i just did each function at different times and wasn't paying attention